dsm/common.go

```diff
@@ -263,4 +264,91 @@ func undoTFstate(param_type string, d *schema.ResourceData) diag.Diagnostics {
 		  Detail:   fmt.Sprintf("[E]: API: PATCH crypto/v1/keys: %s cannot change on update. Please retain it to old value: %s -> %s", param_type, old_val, new_val),
 	})
 	return diags
-}
\ No newline at end of file
+}
+
+// expiry_date create and update
+func parseTimeToDSM(expiry_date string) (string, diag.Diagnostics){
+	layoutRFC := "2006-01-02T15:04:05Z"
+	layoutDSM := "20060102T150405Z"
+	ddate, newerr := time.Parse(layoutRFC, expiry_date)
+	if newerr != nil {
+		return "", diag.FromErr(newerr)
+	}
+	return ddate.Format(layoutDSM), nil
+}
+
+/* Set the key_ops in tf state.
+This function is required, request key_ops order and response key_ops order mighet differ.
+Hence, during `terraform plan`, if there are no changes in key_ops, it should not show any changes.
+*/
+func setKeyOpsTfState(d *schema.ResourceData, key_ops interface{}) diag.Diagnostics{
+	is_same_key_ops := compTwoArrays(key_ops, d.Get("key_ops"))
+	if is_same_key_ops {
+		if err := d.Set("key_ops", d.Get("key_ops")); err != nil {
+			return diag.FromErr(err)
+		}
+	} else {
+		if err := d.Set("key_ops", key_ops); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	return nil
+}
+
+/*
+A BYOK security object can be deleted only when it is in Destroyed state.
+*/
+func deleteBYOKDestroyedSobject(d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	/*
+	BYOK object can be deleted only after destruction.
+	*/
+	error_summary := "[DSM SDK] Unable to call DSM provider API client"
+	if d.Get("state").(string) == "Destroyed" {
+		_, statuscode, err := m.(*api_client).APICall("DELETE", fmt.Sprintf("crypto/v1/keys/%s", d.Id()))
+		if (err != nil) && (statuscode != 204) {
+		    return invokeErrorDiagsWithSummary(error_summary, fmt.Sprintf("[E]: API: DELETE crypto/v1/keys: %v", err))
+		}
+		d.SetId("")
+	} else {
+		return invokeErrorDiagsWithSummary(error_summary, fmt.Sprintf("[E]: API: DELETE crypto/v1/keys: cannot be deleted as security object is in the state %s", d.Get("state").(string)))
+	}
+	return nil
+}
+
+/*
+Throw a warning, if schedule_deletion fails after create.
+*/
+func scheduleDeletionWarning(d *schema.ResourceData, err diag.Diagnostics) diag.Diagnostics {
```

Comment:
This seems like we're making this function tied to the creation of the key... we should pass the `error_summary` as a function parameter so it can be re-used in other contexts scheduling deletion. Alternatively we could just rename the function to something like `scheduledDeletionAfterCreationWarning`.
- I prefer the idea of passing the error summary but will let you decide whether it should be changed at all.

Reply:
This comment is no more required. Hence deleted.